app/src/main/java/net/squanchy/schedule/ScheduleService.java

```diff
@@ -24,33 +29,55 @@
+        Observable<FirebaseSchedule> sessionsObservable = repository.sessions();
+        Observable<FirebaseSpeakers> speakersObservable = repository.speakers();
 
         return Observable.combineLatest(
-                eventObservable,
+                sessionsObservable,
                 speakersObservable,
-                (eventHolder, speakerHolder) -> {
-
-                    List<SchedulePage> pages = map(eventHolder.days, day -> SchedulePage.create(
-                            day.date,
-                            map(day.events, firebaseEvent -> Event.create(
-                                    firebaseEvent.eventId,
-                                    eventHolder.days.indexOf(day),      // TODO do this less crappily
-                                    firebaseEvent.name,
-                                    firebaseEvent.place,
-                                    firebaseEvent.experienceLevel,
-                                    map(
-                                            map(firebaseEvent.speakers, speakerId -> find(
-                                                    speakerHolder.speakers, speaker -> speaker.speakerId.equals(speakerId))
-                                            ),
-                                            speaker -> speaker == null ? "" : speaker.firstName + " " + speaker.lastName
-                                    ))
-                            )
-                    ));
-
-                    return Schedule.create(pages);
-                }
+                composeIntoSchedule()
         ).subscribeOn(Schedulers.io());
     }
+
+    private BiFunction<FirebaseSchedule, FirebaseSpeakers, Schedule> composeIntoSchedule() {
+        return (apiSchedule, apiSpeakers) -> {
+            List<SchedulePage> pages = map(apiSchedule.days, toSchedulePage(apiSchedule, apiSpeakers));
+            return Schedule.create(pages);
+        };
+    }
+
+    private Lists.Function<FirebaseDay, SchedulePage> toSchedulePage(FirebaseSchedule apiSchedule, FirebaseSpeakers apiSpeakers) {
+        return apiDay -> {
+            int dayId = apiSchedule.days.indexOf(apiDay);
+            return SchedulePage.create(
+                    apiDay.date,
+                    map(apiDay.events, toEvent(apiSpeakers, dayId))
+            );
+        };
+    }
+
+    private Lists.Function<FirebaseEvent, Event> toEvent(FirebaseSpeakers apiSpeakers, int dayId) {
+        return apiEvent -> {
+            List<FirebaseSpeaker> speakers = speakersForEvent(apiEvent, apiSpeakers);
+            return Event.create(
+                    apiEvent.eventId,
+                    dayId,      // TODO do this less crappily
+                    apiEvent.name,
+                    apiEvent.place,
+                    apiEvent.experienceLevel,
+                    map(speakers, toSpeakerName()));
+        };
+    }
+
+    private List<FirebaseSpeaker> speakersForEvent(FirebaseEvent apiEvent, FirebaseSpeakers apiSpeakers) {
```

Comment:
Maybe, point is, where? I'd extract when we have a clearer structure and a place where they actually belong.